[Quarkus 2.4.0] Messaging tests fail with ambiguous dependencies for `org.apache.camel.quarkus.component.messaging.it.util.scheme.ComponentScheme`

Maybe something changed recently in Quarkus that has started causing this error.

What confuses me is that the nightly Quarkus platform builds with 999-SNAPSHOT & CQ 2.3.0 are passing. I'd expect to see the same failure.

```
2021-10-11T03:31:18.4927883Z java.lang.RuntimeException: 
2021-10-11T03:31:18.4935938Z java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
2021-10-11T03:31:18.4948543Z 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: Found 14 deployment problems: 
2021-10-11T03:31:18.4960185Z [1] Ambiguous dependencies for type org.apache.camel.quarkus.component.messaging.it.util.scheme.ComponentScheme and qualifiers [@Default]
2021-10-11T03:31:18.4970629Z 	- java member: org.apache.camel.quarkus.component.messaging.it.MessagingCommonResource#componentScheme
2021-10-11T03:31:18.4989377Z 	- declared on CLASS bean [types=[org.apache.camel.quarkus.component.messaging.it.MessagingCommonResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.apache.camel.quarkus.component.messaging.it.MessagingCommonResource]
2021-10-11T03:31:18.4994409Z 	- available beans:
2021-10-11T03:31:18.5106045Z 		- PRODUCER METHOD bean [types=[org.apache.camel.quarkus.component.messaging.it.util.scheme.ComponentScheme, java.lang.Object], qualifiers=[@Default, @Any], target=org.apache.camel.quarkus.component.messaging.it.util.scheme.ComponentScheme getMessagingComponentScheme(org.apache.camel.CamelContext camelContext), declaringBean=org.apache.camel.quarkus.component.messaging.it.util.scheme.MessagingComponentSchemeProducer]
2021-10-11T03:31:18.5142671Z 		- PRODUCER METHOD bean [types=[org.apache.camel.quarkus.component.messaging.it.util.scheme.ComponentScheme, java.lang.Object], qualifiers=[@Default, @Any], target=org.apache.camel.quarkus.component.messaging.it.util.scheme.ComponentScheme getMessagingComponentScheme(org.apache.camel.CamelContext camelContext), declaringBean=org.apache.camel.quarkus.component.messaging.it.util.scheme.MessagingComponentSchemeProducer]
```

Hm, so I built the project with `-Denforcer.skip` and got `[ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.6:yarn (antora-xref-check) on project camel-quarkus-docs` but still I was able to run the tests for `jms-artemis-client` and from the debug log I can see that the bean classes, e.g. `org.apache.camel.quarkus.component.messaging.it.util.scheme.MessagingComponentSchemeProducer`, are **found twice** in the bean archive index. Which could mean that an `ApplicationArchive` is built twice from a dependency or there are two dependencies that contain the given class.

--- a/core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java
@@ -167,7 +167,6 @@ public class ApplicationArchiveBuildStep {
                         GACTV.TYPE_JAR);
                 final ResolvedDependency artifact = userMap.get(key);
                 if (artifact == null) {
-                    userMap.keySet().forEach(k -> System.out.println(" - " + k.getClass().getSimpleName() + " " + k));
                     throw new RuntimeException(
                             "Could not resolve artifact " + key + " among the runtime dependencies of the application");
                 }

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentInjectingDependencyVisitor.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentInjectingDependencyVisitor.java
@@ -640,27 +640,29 @@ public class DeploymentInjectingDependencyVisitor {
     public static PathCollection getResolvedPaths(Artifact artifact, WorkspaceModule module, boolean preferWorkspacePaths) {
         if (preferWorkspacePaths && module != null) {
             final PathList.Builder pathBuilder = PathList.builder();
-            for (ProcessedSources src : module.getMainSources()) {
-                if (src.getDestinationDir().exists()) {
-                    final Path p = src.getDestinationDir().toPath();
-                    if (!pathBuilder.contains(p)) {
-                        pathBuilder.add(p);
-                    }
-                }
-            }
-            for (ProcessedSources src : module.getMainResources()) {
-                if (src.getDestinationDir().exists()) {
-                    final Path p = src.getDestinationDir().toPath();
-                    if (!pathBuilder.contains(p)) {
-                        pathBuilder.add(p);
-                    }
-                }
+            if ("tests".equals(artifact.getClassifier())) {
+                collectResolvedPaths(pathBuilder, module.getTestSources());
+                collectResolvedPaths(pathBuilder, module.getTestResources());
+            } else {
+                collectResolvedPaths(pathBuilder, module.getMainSources());
+                collectResolvedPaths(pathBuilder, module.getMainResources());
             }
             if (!pathBuilder.isEmpty()) {
                 return pathBuilder.build();
             }
         }
         return artifact.getFile() == null ? PathList.empty() : PathList.of(artifact.getFile().toPath());
+    }
+
+    private static void collectResolvedPaths(final PathList.Builder pathBuilder, Collection<ProcessedSources> srcs) {
+        for (ProcessedSources src : srcs) {
+            if (src.getDestinationDir().exists()) {
+                final Path p = src.getDestinationDir().toPath();
+                if (!pathBuilder.contains(p)) {
+                    pathBuilder.add(p);
+                }
+            }
+        }
     }
 
     private static String toGactv(Artifact a) {

